Update to bindgen 0.30.0

anxiousmodernman opened the pull request:
The --no-unstable-rust flag was removed was removed in bindgen 0.26.x and
replaced with an --unstable-rust flag. Now we do not need to provide any flag
to avoid generating unstable rust.

anxiousmodernman committed:
Remove defunct bindgen argument

jamesmunns commented:
Hey @anxiousmodernman, thanks for the PR, sorry it fell through the cracks!

Which version of Bindgen did this change with? Could you please update the documentation to reflect which version of Bindgen you are talking about?

If you update the docs (README.md and SETUP.md, maybe others), I'd be happy to merge!

anxiousmodernman commented:
Bindgen's moved along a bit. I will need to check this against latest to see the behavior.

anxiousmodernman commented:
Some investigation into these cli flags:
* [Argument parser in bindgen 0.25.5](https://github.com/rust-lang-nursery/rust-bindgen/blob/7faa855d01e114c1adc0abd5efd547ed40543882/src/options.rs#L138)
* [...and in bindgen 0.26.0 when it changed](https://github.com/rust-lang-nursery/rust-bindgen/blob/8470e221f9fa5d1f195339737fdd3a7b433cad51/src/options.rs#L138)
* [...and in the latest 0.30.0](https://github.com/rust-lang-nursery/rust-bindgen/blob/05caa820010d44b601b340228d4f95d81958a270/src/options.rs#L164)

I'm going to build with the latest bindgen. If it checks out, I'll update the docs to reflect 0.30.0

anxiousmodernman commented:
With bindgen 0.30.0, "blinky" built fine, but "ble_app_template" gave me errors.

![as_ref_as_mut](https://user-images.githubusercontent.com/396004/30508397-34e0150c-9a4b-11e7-876a-23f6e9ade630.png)

This may be a difference in generated code, or it may be something with the rust-src/component/toolchain It seems like these are standard library traits, so perhaps I didn't get that downloaded correctly.

anxiousmodernman committed:
Upgrade to bindgen 0.30.0

Changes to ble_app_template are required. Various C struct fields are
read and set directly, rather than through `as_ref` and `as_mut`.

anxiousmodernman commented:
I've tested this on a couple of machines now. The latest bindgen breaks the example for bindgen 0.30.0 on the ble_app_template example.

The problem is this type is no longer generated, and the compiler complains about the missing methods `as_ref` and `as_mut`.

```rust
/* present in code generated by bindgen 0.26.3 and earlier, missing in 0.30.0 */
#[repr(C)]
pub struct __BindgenUnionField<T>(::core::marker::PhantomData<T>);
impl <T> __BindgenUnionField<T> {
    #[inline]
    pub fn new() -> Self { __BindgenUnionField(::core::marker::PhantomData) }
    #[inline]
    pub unsafe fn as_ref(&self) -> &T { ::core::mem::transmute(self) }
    #[inline]
    pub unsafe fn as_mut(&mut self) -> &mut T { ::core::mem::transmute(self) }
}
```

The new 0.30.0 generated code uses the bleeding edge-ish contextual keyword `union`. With a few changes, I can get the example to compile. The fields on the generated struct can be [get and set directly](https://github.com/jamesmunns/nrf52dk-sys/pull/4/commits/7739fdad28fadee625bc2d49b2ee36c5ab6e9f39#diff-dd6b519b10fdaafbf84af31451787fa1).

I've only built the code, but not flashed it to a dev board. I'll try that next. 

Refs https://github.com/rust-lang-nursery/rust-bindgen/pull/892

anxiousmodernman commented:
@jamesmunns ready for review: I've tested the bluetooth and blinky example on the dev board using this branch and bindgen 0.30.0, flashed it to the dev board, and things seem normal.

jamesmunns committed:
Update Dockerfile to match documentation

jamesmunns commented:
Looks good to me! Thanks so much for following up with this. I noticed the Dockerfile needed to be updated, so I pushed a quick patch for that as well.
